### PR TITLE
AppGW and AZs

### PR DESCRIPTION
**Pull request by techbunny**
Swapped out LB with AppGW.  Also added support for deploying either an Availability Zone or AVSet based on the region selected.  When deploying to an AVSet, if the VM is set to be single instance, the AVSet won't be used.

### CODE REVIEW

File: Peoplesoft/provider.tf [hunk 1]
```diff
@@ -2,12 +2,11 @@
...
-    version= ">=1.21.0"
+    version= ">=1.28.0"
     tenant_id       = "${var.tenant_id}"
     subscription_id = "${var.subscription_id}"
-    #TODO: for testing, leaving these out ... easier to sign in via CLI.
-  #  client_id       = "${var.client_id}"
-  #  client_secret   = "${var.client_secret}"   
+    client_id       = "${var.client_id}"
+    client_secret   = "${var.client_secret}"   
 }
```
Comment (romitgirdhar) [inline Peoplesoft/provider.tf]: Can we add "partner_id" for PSFT: "1ec53238-c562-5ff9-89f6-558763be3779"

File: Peoplesoft/modules/network/subnets/subnets.tf [hunk 2]
```diff
@@ -8,17 +8,41 @@ resource "azurerm_subnet" "subnet" {
   virtual_network_name = "${var.vnet_name}"
   address_prefix       = "${element(values(var.subnet_cidr_map),count.index)}"
   count = "${length(var.subnet_cidr_map)}"
+  network_security_group_id = "${element(values(var.nsg_ids),count.index)}"
+}
+
+locals {
+    # a temp map to avoid a dependency between the order in which subnets show up in
+    # azurerm_subnet.subnet.*.id and the keys in subnet_cidr_map/nsg_ids.
+    subnetNameToID = "${zipmap(keys(var.subnet_cidr_map),azurerm_subnet.subnet.*.id)}"
```
Comment (romitgirdhar) [inline Peoplesoft/modules/network/subnets/subnets.tf]: What is this for again? It's unclear...

File: Peoplesoft/variables.tf [hunk 3]
```diff
@@ -34,6 +46,7 @@ variable "create_data_disk" {
   default = true
 }
 
+
 # Host Name Prefixes
 variable "compute_hostname_prefix" {
```
Comment (romitgirdhar) [inline Peoplesoft/variables.tf]: Let's hardcode this. No need to have a variable for this.

File: Peoplesoft/variables.tf [hunk 4]
```diff
@@ -44,7 +57,7 @@ variable "compute_hostname_prefix_app" {
   default = "app"
 }
 variable "compute_hostname_prefix_bastion" {
```
Comment (romitgirdhar) [inline Peoplesoft/variables.tf]: Let's hardcode this. No need to have a variable for this.

File: Peoplesoft/variables.tf [hunk 5]
```diff
@@ -58,17 +71,21 @@ variable "compute_hostname_prefix_es" {
   default = "es"
 }
 
-
 variable "compute_hostname_prefix_ps" {
```
Comment (romitgirdhar) [inline Peoplesoft/variables.tf]: Let's hardcode this. No need to have a variable for this.

File: Peoplesoft/variables.tf [hunk 6]
```diff
@@ -58,17 +71,21 @@ variable "compute_hostname_prefix_es" {
...
-  description = "Elastic Search server host resource prefix"
+  description = "Process server host resource prefix"
   default = "ps"
 }
 
 variable "compute_hostname_prefix_tc" {
-  description = "Elastic Search server host resource prefix"
+  description = "Tools Client server host resource prefix"
```
Comment (romitgirdhar) [inline Peoplesoft/variables.tf]: Let's hardcode this. No need to have a variable for this.

File: Peoplesoft/variables.tf [hunk 7]
```diff
@@ -89,19 +106,26 @@ variable "elastic_instance_count" {
   default = 2
 }
 variable "prosched_instance_count" {
-  description = "elastic instance count"
+  description = "prosched instance count"
```
Comment (romitgirdhar) [inline Peoplesoft/variables.tf]: What is prosched?

File: Peoplesoft/variables.tf [hunk 8]
```diff
@@ -58,17 +71,21 @@ variable "compute_hostname_prefix_es" {
...
 }
 
+variable "compute_hostname_prefix_id" {
+  description = "Identity server host resource prefix"
+  default = "id"
+}
+
 # Set instance counts. 
 #Min should be two for Application, Webserver, Elastic and ProSched due to Availablity Set usage.
 variable "compute_instance_count" {
```
Comment (romitgirdhar) [inline Peoplesoft/variables.tf]: Let's ask for only 1 count and remove the rest similar to what we're doing with the VM size

File: Retail/main.tf [hunk 9]
```diff
@@ -7,12 +7,14 @@ locals {
     subnetPrefixes = {
         application     = "${cidrsubnet(var.vnet_cidr, local.vnet_cidr_increase, 1)}"
         bastion-ftp     = "${cidrsubnet(var.vnet_cidr, local.vnet_cidr_increase, 2)}"
-        # Note that ExpressRoute setup needs exactly "GatewaySubnet" as the gateway subnet name.
         AppGWSubnet   = "${cidrsubnet(var.vnet_cidr, local.vnet_cidr_increase, 9)}"
```
Comment (romitgirdhar) [inline Retail/main.tf]: Need to add identity

File: Retail/main.tf [hunk 10]
```diff
@@ -7,12 +7,14 @@ locals {
...
     }
 
-
+    vnet_name = "${var.vnet_cidr == "0" ? 
+    element(concat(data.azurerm_virtual_network.primary_vnet.*.name, list("")), 0) :
+    element(concat(azurerm_virtual_network.primary_vnet.*.name, list("")), 0)}"
+ 
+    vnet_cidr = "${var.vnet_cidr == "0" ? element(concat(data.azurerm_virtual_network.primary_vnet.*.address_space, list("")), 0) : var.vnet_cidr}"
     #####################
     ## NSGs
```
Comment (romitgirdhar) [inline Retail/main.tf]: missing NSG for identity

File: Retail/main.tf [hunk 11]
```diff
@@ -97,7 +123,7 @@ module "create_networkSGsForBastion" {
 module "create_networkSGsForApplication" {
```
Comment (romitgirdhar) [inline Retail/main.tf]: missing module for Identity

File: Retail/main.tf [hunk 12]
```diff
@@ -280,9 +309,8 @@ module "create_idm" {
   create_av_set             = true
```
Comment (romitgirdhar) [inline Retail/main.tf]: need to remove this variable and what we did for PSFT

File: Peoplesoft/provider.tf (hunk 1 above)
Comment (romitgirdhar) [reply Peoplesoft/provider.tf]: Can we also do the same for Retail: "partner_id" = "e7a6fb4f-fce6-57a3-abfc-c6dfafd44692"

File: Peoplesoft/main.tf [hunk 13]
```diff
@@ -77,21 +84,27 @@ locals {
             source_port_ranges =  "*" 
             source_address_prefix = "${local.subnetPrefixes["bastion"]}"                
             destination_port_ranges = "22" 
-            destination_address_prefix = "*"             
+            destination_address_prefix = "${local.subnetPrefixes["webserver"]}"
```
[the remark this answers is not in the document]
Comment (techbunny) [reply Peoplesoft/main.tf]: It means that the bastion is allowed to talk to the webserver.

File: Peoplesoft/main.tf [hunk 14]
```diff
@@ -77,21 +84,27 @@ locals {
...
-            destination_address_prefix = "*"             
+            destination_address_prefix = "${local.subnetPrefixes["webserver"]}"               
         },
                {
             source_port_ranges =  "*" 
             source_address_prefix = "${local.subnetPrefixes["elasticsearch"]}"                
-            destination_port_ranges = "*" 
-            destination_address_prefix = "*"             
+            destination_port_ranges = "80" 
+            destination_address_prefix = "${local.subnetPrefixes["webserver"]}"
```
[the remark this answers is not in the document]
Comment (techbunny) [reply Peoplesoft/main.tf]: These are the inbound rules... so webserver accepts traffic from bastion, elasticsearch and the load balancer.

File: Peoplesoft/modules/network/subnets/subnets.tf (hunk 2 above)
Comment (techbunny) [reply Peoplesoft/modules/network/subnets/subnets.tf]: Its to help make sure the subnets get matched up with the right NSGs.  It was a shawn thing that existed in Retail but was missing in Peoplesoft and I wasn't sure if it was related to subnet to NSG issue I was having earlier.

File: Peoplesoft/variables.tf (hunk 7 above)
Comment (techbunny) [reply Peoplesoft/variables.tf]: Process Scheduler VMS

File: Peoplesoft/modules/network/subnets/subnets.tf (hunk 2 above)
Comment (romitgirdhar) [reply Peoplesoft/modules/network/subnets/subnets.tf]: i'm not sure we need this, tbh. I think ordering is taken care of. But, it doesn't look like it's doing any harm, so, we can keep it for now

**Pull request merged**